tests/functional/autoprompt/test_prompttoolkit.py

```diff
@@ -51,45 +58,129 @@ def run(self, pre_run):
         pre_run()
 
 
-class TestPromptToolkitPrompterBuffer(unittest.TestCase):
-    """This set of tests tests that we set the buffer (aka "construction zone")
-    correctly. Some of these tests test against specific edge cases that have
-    previously been known to produce unexpected behavior.
+class FakeInput(DummyInput):
 
-    """
+    def fileno(self):
+        return 0
+
+    @property
+    def closed(self):
+        return False
+
+
+class ApplicationStubber:
+    _KEYPRESS = 'KEYPRESS'
+    _ASSERTION = 'ASSERTION'
+
+    def __init__(self, app):
+        self._app = app
+        self._app.input = FakeInput()
+        self._queue = []
+        self._assertion_callback = lambda x: True
+        self._default_keybindings = load_key_bindings()
+
+    def add_key_assert(self, key=None, assertion=None):
```

Comment:
Usually key press and assertion come together if I split them I'll have to pass key that was pressed on the previous step with assertion to be able to give meaningful error message

Reply:
That's fair. Let's instead do the following:
1. Add a `add_keypress(self, key)` method just for adding a key press action
2. Rename this method to `add_keypress_with_assertion` and make the `key` and `assertion` required.

I'm mainly suggesting this because the method interface is a little hard to follow given there are two ways to use it (listed above) and it seems like it would make a little easier to follow if we just broke it up into two methods.